### PR TITLE
fix(messaging, android): Replace deprecated AsyncTask API and other deprecated API

### PR DESCRIPTION
**Pull request by russellwheatley**
### Done
- [x] Replaced deprecated AsyncTask.
- [x] Replaced deprecated getParcelableExtra() and replaced with passing Parcel through Intent for background/quit app states.
- [x] Removed some unused code.

I tested messages received in open/background/quit and also when opening app with message. Working as intended.

## Related Issues

Part of: https://github.com/firebase/flutterfire/issues/8073

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [X] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingBackgroundExecutor.java [hunk 1]
```diff
@@ -227,29 +228,38 @@ public void notImplemented() {
             }
           };
     }
+    // RemoteMessage is passed as byte array. Check it exists first
+    byte[] parcelBytes =
+        intent.getByteArrayExtra(FlutterFirebaseMessagingUtils.EXTRA_REMOTE_MESSAGE);
+    if (parcelBytes != null) {
+      Parcel parcel = Parcel.obtain();
```
Comment (Lyokone) [inline packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingBackgroundExecutor.java:235]: This cannot fail? Or it should be part of the try/catch?

File: packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/JobIntentService.java [hunk 2]
```diff
@@ -342,32 +344,42 @@ public void complete() {
...
-    @Override
-    protected void onCancelled(Void aVoid) {
-      processorFinished();
+  final class CommandProcessor {
+    private final Executor executor = Executors.newSingleThreadExecutor(); // Background thread
+    private final Handler handler = new Handler(Looper.getMainLooper()); // UI Thread
+
+    public void execute() {
+      executor.execute(
+          new Runnable() {
```
Comment (Lyokone) [inline packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/JobIntentService.java:353]: Should we store this runnable to cancel it manually inside `cancel()` if needed?

File: packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java [hunk 3]
```diff
@@ -613,4 +614,10 @@ public Task<Void> didReinitializeFirebaseCore() {
 
     return taskCompletionSource.getTask();
   }
+
+  private Map<String, Object> safelyCastToMap(Object obj) {
+    @SuppressWarnings("unchecked")
+    Map<String, Object> result = (Map<String, Object>) obj;
+    return result;
+  }
```
Comment (TarekkMA) [inline packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java:622]: Is this here just to get rid of the warning? if so shouldn't it be renamed to `uncheckedCastToMap`?

File: packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingBackgroundExecutor.java (hunk 1 above)
Comment (russellwheatley) [reply packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingBackgroundExecutor.java:235]: It should never fail.

File: packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/JobIntentService.java (hunk 2 above)
Comment (russellwheatley) [reply packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/JobIntentService.java:353]: I don't think it is necessary, it is a like for like replacement replacing deprecated AsyncTask. The execute(Runnable) is running in the background thread (like `doInBackground` for AyncTask). When that work is complete, it calls the `processorFinished()` finished on the main thread (like `onPostExecute` for AsyncTask). See documentation: https://developer.android.com/reference/android/os/AsyncTask#the-4-steps

File: packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java (hunk 3 above)
Comment (russellwheatley) [reply packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java:622]: Yeah, unfortunately, SuppressWarnings are only allowed in certain contexts. I had to take that cast out in order to suppress it. I only suppress warnings on API that I believe to be stable (i.e. no occurrences of failure).

Comment (TarekkMA) [reply packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java:622]: but we need to change it to `uncheckedCastToMap` for example, to give the intention that we know that's safe to cast it unchecked. what do you think?

File: packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/JobIntentService.java (hunk 2 above)
Comment (russellwheatley) [reply packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/JobIntentService.java:353]: To add to that, the only time cancel is called using previous AsyncTask, we also call cancel() here using latest implementation: https://github.com/firebase/flutterfire/pull/12580/files#diff-3d9ef1ece89413638d5bc6907c220242d3b2f489db5900782615c504073437f1R561

**Review by TarekkMA — APPROVED** [review]

**Review by Lyokone — APPROVED** [review]

**Pull request merged**